sections/performance/block-loop.md

```diff
@@ -0,0 +1,57 @@
+Node handles the Event Loop on a single thread rotating through multiple queues. Operations with high complexity, large json parsing, unsafe regex querieies, and long awaited promises are some of the operations that can cause the event loop to stall. Preventing this is as easy as setting up monitoring for the the Event Loop and resolving the blocker.
```

Comment:
"handles the Event Loop on a single thread " -> "mostly on a single thread" (just to be precise)

"Operations with high CPU cost" (it's not about complexity, IO can be too complex)

"querieies" - typo?

"long awaited promises" - what do you mean here? not sure this is in-general true, usually promises are used to wait for IO

"Preventing this is as easy as setting up monitoring for the the Event Loop and resolving the blocker." - monitoring is detection, not prevention. Let's keep the monitoring to a dedicated bullet. My thoughts "Avoid this off-loading CPU intensive tasks to dedicated workload that doesn't serer users (e.g. job server), or by breaking long tasks into small steps that stand on the event queue lines, by using crypto libraries that utilize the event loop thread pool. To name a few techniques that can keep the event loop alive"

Reply:
A lot of my info comes from this article: https://nodejs.org/en/docs/guides/dont-block-the-event-loop/

From this section: https://nodejs.org/en/docs/guides/dont-block-the-event-loop/#don-t-block-the-event-loop

For promises: "You should make sure you never block the Event Loop. In other words, each of your JavaScript callbacks should complete quickly. This of course also applies to your await's, your Promise.then's, and so on."

For computational complexity
"A good way to ensure this is to reason about the "computational complexity" of your callbacks."

Also, I thought IO is going to be a separate bullet point. IO blocks the event loop for a different reason: 
```
I/O-intensive
    DNS: dns.lookup(), dns.lookupService().
    File System: All file system APIs except fs.FSWatcher() and those that are explicitly synchronous use libuv's threadpool.
```
If not, can you give me a counter example?

Reply:
> For promises: "You should make sure you never block the Event Loop. In other words, each of your JavaScript callbacks should complete quickly. This of course also applies to your await's, your Promise.then's, and so on."

Note there is a signficant difference between what you wrote and the article - you alluded to 'long awaited promises', they refer to the .then part. Practically, most promises wrapping some IO work, if you wait for long time then it's not related to CPU work. Their advice of not slowing down the .then handler makes sense, but it's too specific, we can generalize this advice to any other code...

> "A good way to ensure this is to reason about the "computational complexity" of your callbacks."

Well, it's probably right technically, is it a simple and well understood term? I guess the answer is subjective. I suggest that our guide should curate and simplify those technical guides

> For non-sync IO, should I state that you should also watch for large IO tasks?
 
I think it doesn't belong here
 
 

Note that I also included in my comment multiple mitigation techniques

Reply:
I'll definitely include many of your suggestions. I'm just trying to hone in on "high CPU usage" because it is super generic. I want to tell people what to look for.

I believe that these are three examples: `Operations with high complexity, large json parsing, unsafe regex queries`. So "high complexity" is just one example.

Also, promises are also used a lot for db access. If you have long queries that you are are `awaiting` you could block the loop like our example, I think.

Anyway, I'm just trying to give devs examples of "cpu work" to look for :D. I'll get your edits in this weekend.

Reply:
OKay, after reading this I'd retract my comment about `awaiting`. I was missing the idea that promises are added to the job queue: https://github.com/getify/You-Dont-Know-JS/blob/master/async%20%26%20performance/ch1.md. Soooo sorry about that :P

Reply:
Yes, if you await on IO operation (E.g. DB call) - this will not the event loop